common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -118,76 +119,433 @@ public Collation(
+          String localeUppercase = localeName.toUpperCase();
+          // locale names are unique case-insensitively
+          assert (!ICULocaleMapUppercase.containsKey(localeUppercase));
+          ICULocaleMapUppercase.put(localeUppercase, localeName);
+        }
+        ICULocaleNames = ICULocaleMap.keySet().toArray(new String[0]);
+        Arrays.sort(ICULocaleNames);
+        // maximum number of locale ids as defined by binary layout
+        assert (ICULocaleNames.length <= (1 << 12));
+        for (int i = 0; i < ICULocaleNames.length; ++i) {
+          ICULocaleToId.put(ICULocaleNames[i], i);
+        }
+      }
+
+      private static final int UNICODE_COLLATION_ID =
+        new CollationSpecICU("UNICODE", CaseSensitivity.CS, AccentSensitivity.AS).collationId;
+      private static final int UNICODE_CI_COLLATION_ID =
+        new CollationSpecICU("UNICODE", CaseSensitivity.CI, AccentSensitivity.AS).collationId;
+
+      private final CaseSensitivity caseSensitivity;
+      private final AccentSensitivity accentSensitivity;
+      private final String locale;
+      private final int collationId;
+
+      private CollationSpecICU(String locale, CaseSensitivity caseSensitivity,
+          AccentSensitivity accentSensitivity) {
+        this.locale = locale;
+        this.caseSensitivity = caseSensitivity;
+        this.accentSensitivity = accentSensitivity;
+        int collationId = ICULocaleToId.get(locale);
+        collationId = SpecifierUtils.setSpecValue(collationId, IMPLEMENTATION_PROVIDER_OFFSET,
+          ImplementationProvider.ICU);
+        collationId = SpecifierUtils.setSpecValue(collationId, CASE_SENSITIVITY_OFFSET,
+          caseSensitivity);
+        collationId = SpecifierUtils.setSpecValue(collationId, ACCENT_SENSITIVITY_OFFSET,
+          accentSensitivity);
+        this.collationId = collationId;
+      }
+
+      private static int collationNameToId(
+          String originalName, String collationName) throws SparkException {
+        // search for the longest locale match because specifiers are designed to be different from
+        // script tag and country code, meaning the only valid locale name match can be
+        // the longest one
+        int lastPos = -1;
+        for (int i = 1; i <= collationName.length(); i++) {
+          String localeName = collationName.substring(0, i);
+          if (ICULocaleMapUppercase.containsKey(localeName)) {
+            lastPos = i;
+          }
+        }
+        if (lastPos == -1) {
+          throw collationInvalidNameException(originalName);
+        } else {
+          String locale = collationName.substring(0, lastPos);
+          int collationId = ICULocaleToId.get(ICULocaleMapUppercase.get(locale));
+
+          // try all combinations of AS/AI and CS/CI
+          CaseSensitivity caseSensitivity;
+          AccentSensitivity accentSensitivity;
+          if (collationName.equals(locale) ||
+              collationName.equals(locale + "_AS") ||
```

Comment:
Is `collationName` expected to be uppercase?

Reply:
The name is not expected to be of a particular case, we convert it to uppercase [here](https://github.com/apache/spark/pull/46180/files#diff-640c14aa5d7473df79b2435ce5a327dffcc16ca29354b153956b4f8d19fdb16cR278) before performing comparisons.